### PR TITLE
Preserve Extern modifier when running code cleanup

### LINKED ISSUE #29561: "Add accessibility modifiers" C# formatting rule breaks code
We do not use &quot;private&quot; and &quot;internal&quot; everywhere because they are default modifiers and we don't want to see defaults in our code. That rule adds them, making lots of diffs for no reason. This rule should not be opt-out, it should be opt-in.

Another problem is methods with &quot;extern&quot; modifier. This rule removes &quot;extern&quot;, breaking our interop code. Please don't do that.

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/293507/add-accessibility-modifiers-c-formatting-rule-brea.html
VSTS ticketId: 648040_
_These are the original issue comments:_

Jinu Joseph [MSFT] on 8/5/2018, 00:53 AM (22 days ago): <p>Thank you for your feedback! For us to investigate this further, could you please provide code snippet which leads to this issue, specific example will help us to investigate better.  We look forward to hearing from you!</p>

Roy Scheerens on 8/5/2018, 03:28 AM (22 days ago): <p>I have this problem as well, and it seems to happen with every extern function/method that doesn't have an access modifier. If I enable the additional code cleanup, and the 'add accessibility modifiers' option, it will remove extern and replace it with private.</p><p></p><pre>using System;<br>using System.Runtime.InteropServices;<br><br>namespace ConsoleApp8<br>\{<br>  class Program<br>  \{<br>    [DllImport("User32.dll", CharSet = CharSet.Unicode)]<br>     static extern int MessageBox(IntPtr h, string m, string c, int type);<br><br>    static void Main(string[] args)<br>    \{<br>      MessageBox(IntPtr.Zero, "hello world", "My Message Box", 0);<br>    }<br>  }<br>}</pre>
image

Jinu Joseph [MSFT] on 8/5/2018, 10:52 PM (21 days ago): <p>thanks much for the example, we will investigate the issue</p>

_These are the original issue solutions:_
(no solutions)

### PR DESCRIPTION
**Pull request by chborl**
Fixes #29561 

outstanding issues:

- [ ] Fix the formatting of the line after the field declaration

- [ ] Add tests for other modifiers

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs [hunk 1]
```diff
@@ -171,6 +171,31 @@ private void Method()
                 (CodeCleanupOptions.AddAccessibilityModifiers, enabled: true));
         }
 
+        [Fact]
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs]:
```suggestion
        [Fact]
        [WorkItem(29561, "https://github.com/dotnet/roslyn/issues/29561")]
```

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 2]
```diff
@@ -1825,6 +1825,10 @@ private static SyntaxTokenList AsModifierList(Accessibility accessibility, Decla
                 list = list.Add(SyntaxFactory.Token(SyntaxKind.RefKeyword));
             }
 
+            if (modifiers.IsExtern)
+            {
+                list = list.Add(SyntaxFactory.Token(SyntaxKind.ExternKeyword));
+            }
             return list;
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]:
```suggestion

            return list;
```

File: src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs [hunk 3]
```diff
@@ -176,6 +180,7 @@ private enum Modifiers
             Async = 0x0800,
             WriteOnly = 0x1000,
             Ref = 0x2000,
+            Extern = 0x4000
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs]:
```suggestion
            Extern = 0x4000,
```

File: src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs [hunk 4]
```diff
@@ -84,6 +86,8 @@ public static DeclarationModifiers From(ISymbol symbol)
 
         public bool IsAsync => (_modifiers & Modifiers.Async) != 0;
 
+        public bool IsExtern => (_modifiers & Modifiers.Extern) != 0;
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs]: 💡 It would be good to keep this in order (moved after `IsRef`)

File: src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs [hunk 5]
```diff
@@ -26,7 +26,8 @@ private DeclarationModifiers(Modifiers modifiers)
             bool isWithEvents = false,
             bool isPartial = false,
             bool isAsync = false,
-            bool isWriteOnly = false)
+            bool isWriteOnly = false,
+            bool isExtern = false)
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs]: ❓ Is this change needed? It seems like this parameter is unused. I noticed, for example, that there is no `isRef` parameter either.

Comment (sharwell) [inline src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs]: ❓ Does `From(ISymbol)` need to be updated to set `isExtern` appropriately?

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 6]
```diff
@@ -1825,6 +1825,10 @@ private static SyntaxTokenList AsModifierList(Accessibility accessibility, Decla
                 list = list.Add(SyntaxFactory.Token(SyntaxKind.RefKeyword));
             }
 
+            if (modifiers.IsExtern)
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: ❓ Is there a VB equivalent to this?

**Comment by chborl** [comment]
@mavasani and @sharwell , this is related to https://github.com/dotnet/roslyn/pull/32803 which we are going to cover in our next design review.  Should I hold off on this PR until we decide what to do with DeclarationModifiers?
